src/awkward/_v2/operations/ak_fields.py

```diff
@@ -27,4 +29,4 @@ def fields(array):
 
 def _impl(array):
     layout = ak._v2.operations.to_layout(array, allow_record=True, allow_other=False)
-    return layout.fields
+    return copy.copy(layout.fields)
```

Comment:
I'd suggest using `.copy` rather than `copy.copy` here, because we know that we have a list.

Reply:
Done!

Reply:
Oh, I forgot that.

https://github.com/scikit-hep/awkward/blob/1a0858e20cc2af56b2fe01541721f88ac61e2f20/src/awkward/_v2/forms/numpyform.py#L198-L200

Well, the same is true of `parameters`: we know that we have a dict:

https://github.com/scikit-hep/awkward/blob/1a0858e20cc2af56b2fe01541721f88ac61e2f20/src/awkward/_v2/forms/form.py#L259-L263

And a dict also has a `.copy()`. I'll put that in.

Reply:
Taking a `.copy()` of a dict would be a shallow copy w.r.t the values. Is there any chance that we want a deep copy? Are there any uses of `parameters` that involve setting a mutable object e.g. dict, list etc.

Reply:
(So would `copy.copy`: also shallow.)

In principle, `parameters` can be any JSON-able thing. In practice, they've always been strings. The way it looks like things are going, they'll always be strings. If we have some non-string lists and dicts, then in principle we'd have to protect them with `deepcopy`.

```python
>>> parameters = {
...     "__record__": "SomeLongName",
...     "__categorical__": False,
...     "units": "cm",
...     "__doc__": "This is some thing. Wow."
... }
>>> def copy1(what, how_many):
...     for _ in range(how_many):
...         tmp = what.copy()
... 
>>> def copy2(what, how_many):
...     for _ in range(how_many):
...         tmp = copy.copy(what)
... 
>>> def copy3(what, how_many):
...     for _ in range(how_many):
...         tmp = copy.deepcopy(what)
... 
>>> def copy4(what, how_many):
...     for _ in range(how_many):
...         tmp = json.loads(json.dumps(what))
... 
>>> start = time.time(); copy1(parameters, 1000000); time.time() - start
0.05509829521179199
>>> start = time.time(); copy2(parameters, 1000000); time.time() - start
0.18173909187316895
>>> start = time.time(); copy3(parameters, 1000000); time.time() - start
2.2748003005981445
>>> start = time.time(); copy4(parameters, 1000000); time.time() - start
3.071228504180908
```

I don't think the unlikely possibility of weird types justifies the expense. Or wait, maybe

```python
>>> def copy5(what, how_many):
...     for _ in range(how_many):
...         if all(isinstance(x, (str, bool, numbers.Integral, numbers.Real)) for x in what.values()):
...             tmp = what.copy()
...         else:
...             tmp = copy.deepcopy(what)
... 
>>> start = time.time(); copy5(parameters, 1000000); time.time() - start
0.5682976245880127
```

is only a factor-of-10 cost, which goes into the expensive branch on weird cases (none of ours).

Reply:
Oops, I misread this as `copy.deepcopy`.